### DISCUSSION TITLE
Translation of .po language files - howto

### DISCUSSION THREAD

**Discussion by obw-go** [GrandOrgue/grandorgue#1950] [Developers]
It's probably not a bad idea to have a seperate discussion topic about the translation of .po language files in general. 

What are possible approaches and their (dis)advantages?
 
I played around with the nl.po file that's going to be part of the help directory, first trying out different approaches that didn't work well. For now working with [poedit](https://poedit.net/) and a separate AI produced file gohelpnl.docx with the Dutch translations of the help texts seems to work ok, because I can in one step replace all the systematic translation errors in gohelpnl.docx, while still preserving the technical pieces with references to links and pictures. 

Examples of translation errors (in Dutch) occur with homonyms such as organ, key, event, stop, enclosure, etc. Some of these homonyms are only properly translated if they occur in a sentence that contains other organ-related words. Replacing the wrong translation (in Dutch: orgaan, sleutel, gebeurtenis, stop, omsluiting,...) I do in the file gohelpnl.docx.

Now it's a matter of copy paste and improve the (Dutch) translation from gohelpnl.docx to poedit and during this process improve gohelpnl.docx by error-replacements. My impression is that this works pretty fast. In approx two hours I could add 176 of the 2466 ID's in poedit. So doing the complete translation would cost me approx 28 hours. That's workable, doing it an hour a day for a month or so...
Although... the devil is in the details. The fine-tuning and coordination of the translations of both the program language file and the help language file takes roughly the same time. Especially if you are not familiar with the technical details of the organ, GO program, and sample set terms and codes, it is useful during fine-tuning to carefully study the program and help context in which the translations function.

For who is interested in this approach, I add as attachment the English source file for making the AI translation, that was done for free [on this site](https://translate.google.nl/?sl=nl&tl=en&op=docs). Here the English source: [gohelpuk.txt](https://github.com/user-attachments/files/16321575/gohelpuk.txt) I used.

See lower [in this post](https://github.com/GrandOrgue/grandorgue/discussions/1950#discussioncomment-10166424) for a detailed description of the actual implementation of the translations in the GO sources. 

If others do translation work on the .po files, please share your approach(es) here!

**Comment by obw-go** [comment]
If you want to know a bit more about the AI translation problem with homonyms, you can go to the page https://gemini.google.com/app and ask the following questions:


Is there a way to nudge the translation of a document when it's tranlated by "Google Translate", especially regarding homonyms?
The English word organ is the Dutch translation most of the time translated as "orgaan", while it should be "orgel". Is there a way to nudge the translation process?

**Comment by obw-go** [comment]
In the [README-translate file](https://github.com/GrandOrgue/grandorgue/blob/master/README.translate) info is given about howto add new translations and change existing translation files.

This file README.translate assumes that the translator made a setup for adding, changing and testing the translation-files. 

I suppose that most of the time the code-developers are not so keen on working also on the menu- and help-manual-translation job, even in their own tongue. And apart from that: a series of other languages, code-developers are not familiar with, stay open. So some help can come from GrandOrgue users who are not so keen on code-developement and/or sample set construction, but want to contribute to the GO project by working on the translation of the menu- and help-manual-strings.

Most GrandOrgue users work with Windows, so they of course want to do translation  work on their Windows system as well.
Suppose you are in this position and want to contribute to the translation of GrandOrgue in your tongue.
Then it's possible to use under Windows the opensource program [VirtualBox.](https://www.virtualbox.org/wiki/Downloads).

As second step you can install (for now) [Ubuntu 20.4 Desktop version](https://releases.ubuntu.com/focal/) under VirtualBox.

Warning: During the setup of Ubuntu in VirtualBox, set the VDI disksize NOT to the default 10 GB, but raise this number to 20GB (in the advanced settinge tab), otherwise you will get stuck during the GrandOrgue installation processes!

After you installed Ubuntu under VirtualBox, you can follow the steps for the installation of GrandOrgue source files under Ubuntu. 

For this you **DON't** follow the steps as described in BUILD.md, because these steps get the source files straight from the master branche ( read: main directory) of GrandOrgue on GitHub.

Instead you FIRST have to make your own fork (read: kind of copy) of the GrandOrgue source files on GitHub and have those files copied to Ubuntu.

If you do this for the first time, the first steps can be done using FireFox under Ubuntu.
Start FireFox and goto the page https://github.com/GrandOrgue/grandorgue
Click on the first tab on the top left with the label **<> code**
Below the tabs you see buttons on the right side. Click on the  button with the name **Fork** in order to make your own fork that is a combination of your GitHub username with the string "/grandorgue". In this text I will refer to this as **user/grandorgue**
The **master branche** of your fork is a copy of the master branche of the GrandOrgue source files.

Now there are two files to be added and/or translated: a file with strings that appear in the GrandOrgue program (e.g. menu options) and a file with the strings for the help that provides detailed information about the program options.
The first one is in the source directory /po, the second one in the source directory /help. 
In order to make life not too easy for you, the files in those different directories have the same name: xx.po (where xx stands for the two characters of your language, e.g. de for German, fr for French etc.}

It's "best practice" in the GrandOrgue development strategy to place added and changed files in your own fork. But NOT in the master branch of your fork, because this one should always mirror the master branch of GrandOrgue.
So you have to make two extra branches, derived from your own fork, one for each language file: /po/xx.po and /help/xx.po.

For making these two extra branches:
With the tab **<> code** activated in your fork, you see on the left side of the screen a button with the name **master**.
Click on this button and you see a text field with the text "Find or create a branch..."
In this field typ a name for the branch that will be related with the translation of /po/xx.po, e.g. **AddLanguage** (where Language has to be the name of your language) and click on the confirmation button if the name is ok with you.
Now Github immedeately switches to this new branch, so first you have to go back to the **master branch.**
Click on the button with tne name AdLanguage and in the dropdown menu choose **master**
Repeat the same steps for the second extra branche related to the translagion of /help/xx.po. 
For this branche you can choose a name like **AddLanguageHelp** so you know that this branch is about the help file.

Now you are ready for the preparation of the grandorgue setup under Ubuntu.

First of all: install the program poedit under Ubuntu because this program is used for editting the langauge files during the translation process, while protecting the English source strings. You can do that using FireFox and then from the poedit download page.

I give here a shortcut to the **terminal** **commands** that are to be used for that. 
Open the Terminal Window by rightlclicking your mouse on the icon of the home folder and choosing the option 'Open in Terminal'. 
In this Terminal Window type the next commands (written in gray):
.................... command for downloading the GO sources from your fork, where /user/ stands for your GitHub username:
`git clone --recurse-submodules https://github.com/user/grandorgue.git`
Then go from the home directory to the /grandorgue directory that contains the 'master' source files:
`cd grandorgue`

..........now start working in a separate branch for the file /po/po.nl (containing the strings for the program)
.................... command to create a new branch to do the actual work on, e.g. AddLanguage for the /po/xx.po file:
`git checkout -b AddLanguage`
................... commands to make a build directory, go to it and give the build commands:
```
mkdir build
cd build
cmake .. -DCMAKE_BUILD_TYPE=Release
make
```
----------- the last command will take considerable time... given percentages indicate how far it is...

After these preparatory steps, you can follow the instructions for handling the language-files in the [README-translate file](https://github.com/GrandOrgue/grandorgue/blob/master/README.translate)

Let's give a short summary here from that file:
First the commands to update the language file /po/xx.po, e.g. with new strings that have not been translated yet:
```
make update-pot
make merge-pot
```
The next command gives information about what was changed:
`git status`

If git status show files changed that you don't intend to change, use git restore on any of the (.po) files you don't want to be changed/included for your particular branch.

After doing a series of changes, it's always a good idea to check if they work out ok in the program.
For that you can **do a rebuild**.
You could use a shell file for that with the name poupdate2.sh with the next commands in it.
```
cd build
cmake .. -DCMAKE_BUILD_TYPE=Release
make
cd bin
./GrandOrgue
```
Save this file in the directory /grandorgue
In the terminal goto the directory /grandorgue and type:
`./poupdate2.sh`
If the terminal refuses to run this shell file, you first have to give this file these rights by using the next command:
`chmod +x poupdate2.sh`

After running poupdate2.sh the build is made and the GO program starts up so you can check the changes you made.
Just leave the terminal open for typing commands later (minimize it) after closing the GO program.
Continue the work on the file /po/xx.po in the program poedit.
After that you can open the terminal window again, click on the arrow-up-key (whichs show the last used command ./poupdate2.sh) and hit Enter to rebuild again.


After all the edit work has been done on /po/nl.po using poedit and was tested by builds, do the next steps:
make a copy of the file /grandorgue/po/xx.po and put this copy e.g. in /downloads and make another copy with the name containing the present date: xx_YYYYMMDD.po. Copy both files to a separate directory outside the /grandorgue directory, e.g. with the name go_backup. It's also recommended to copy the files to a similar directory on a cloudserver.

.... then use the next command to commit the changes in this branche AddLanguage:
`git commit -a -m "A message explaining what you did."`

Now the file is ready for the **Pull request** on GitHub (this is the request to add the changed file to the GrandOrgue master branch). 
Before this merging of the changed file into the sources, the changed file has to be reviewed and approved by serveral developers.

The Pull request can be done this way:
Start FireFox (under Ubuntu)
Open in FireFox the page https://github.com/**user**/grandorgue/tree/AddLanguage
If this branch AddLanguage does not exist yet in GitHub, add the branch (howto is described earlier in this text),
Open in this branch on the left side the directory po.
Click on the right side on the button `Add file` and choose for `Upload files`
Now drag&drop the changed xx.po from the directory where you saved it.
After that click lower down on the `Commit changes` button. 
Then click in the yellow coloured banner the `Compare & pull request` button.
Explain in the message what was changed in po/xx.po
and click on the green button `Create pull request`.

Now wait for the reviews, about which you will be informed by email.
When you get a confirmation that the updated /po/xx.po was merged,
you can remove the branch AddLanguage.

Be sure that BEFORE a next translation session you first **on the GitHub page**
sync the master of your fork with the GrandOrgue master (so your fork is updated with the most recent GO source files)
For this open in FireFox the page https://github.com/**user**/grandorgue/tree/master
click the button `Sync fork`
After that make in your fork in GitHub the branch AddLanguage again, which will now contain the most recent source files as well.
Replacing the branch by the next one with the same name is also necessary due to the time-stamps GitHub is using to prevent mixing up older files.
Remove a branch in your fork this way:
Goto your fork.
Clik on the button on the left side with the word `master`
and in the drop-down list bottom on the option `View all branches`
In the next windows you can delete the branch.
After that click go back to the master page of your fork,
click again on the button with the word `master`
and make a fresh branch with the name AddLanuage (or AddLanguageHelp).

Be sure that before adding the translation of new strings to a xx.po file in the /grandorgue directory **under Ubuntu**, you first update the ‘master’ with the most recent GO source files from your updated fork master. For that you can use the next commands in a terminal window in the directory /grandorgue:
```
git checkout master
git pull origin master
```

For the translation of the help file /help/xx.po you can follow the same procedure, but some commands are different, e.g. those for updating the strings in /help/xx.po. 
======= todo ========  make help-merge-pot doens't work ok for now, messing up the order of the strings... (meaning: producing a seemingly random order of the strings in terms of xml order, no longer keeping track with the AI translation that was based on the first /help/xx.po without translations and all strings in a (chrono)logical xml order)

**Comment by obw-go** [comment]
@larspalo @oleg68 

Hi Lars & Oleg,
Related to the follow up of the translations, I have two questions.

1 How about the pictures in the Help?
I see pictures taken from the Linux version and pictures from the Windows version.
Is there a preference for this mix of pictures from different operating systems?

2 How often would it be good for the translators to check for new strrings in the .po files and update them?
I could image that this could be related to a next release version of GO.
In that case: How could an upcoming release be communicated to the translators?
wkr, Otto

**Reply by larspalo** [reply]
> 1 How about the pictures in the Help?

As far as I know, there hasn't been any special thoughts about what OS the images should come from, I think that the contributors have just used what they had available at the moment. The intention with GO has always been that it should be cross-platform - which also was one of the historic reasons for using wxWidgets. Thus, from my point of view, it really doesn't matter if the images come from different platforms but they should of course be an accurate depiction of how GO currently renders it.

> 2 How often would it be good for the translators to check for new strrings in the .po files and update them?

Basically when the translator thinks it's needed, in my opinion. When something significant has changed in the source text or an obvious error has been detected that the translator thinks it's worth fixing could be some guidance.

We don't have any fixed release time schedule but there might be indications at https://github.com/GrandOrgue/grandorgue/milestones that could be used to track when a release is imminent...
